plugins/aap-backend/src/providers/AapResourceEntityProvider.ts

```diff
@@ -0,0 +1,170 @@
+      if (!options.schedule && !providerConfig.schedule) {
+        throw new Error(
+          `No schedule provided neither via code nor config for AapResourceEntityProvider:${providerConfig.id}.`,
+        );
+      }
+
+      const taskRunner =
+        options.schedule ??
+        options.scheduler!.createScheduledTaskRunner(providerConfig.schedule!);
+
+      return new AapResourceEntityProvider(
+        providerConfig,
+        options.logger,
+        taskRunner,
+      );
+    });
+  }
+
+  private constructor(
+    config: AapConfig,
+    logger: Logger,
+    taskRunner: TaskRunner,
+  ) {
+    this.env = config.id;
+    this.baseUrl = config.baseUrl;
+    this.authorization = config.authorization;
+    this.owner = config.owner || 'AAP';
+    this.system = config.system || '';
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+
+    this.scheduleFn = this.createScheduleFn(taskRunner);
+  }
+
+  private createScheduleFn(taskRunner: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:run`;
+      return taskRunner.run({
+        id: taskId,
+        fn: async () => {
+          try {
+            await this.run();
+          } catch (error) {
+            this.logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  getProviderName(): string {
+    return `AapResourceEntityProvider:${this.env}`;
+  }
+
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn();
+  }
+
+  async run(): Promise<void> {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    this.logger.info(`Discovering ResourceEntities from AAP ${this.baseUrl}`);
+
+    const entities: Entity[] = [];
+
+    const jobTemplates = await listJobTemplates(
+      this.baseUrl,
+      this.authorization,
+    );
+
+    for (const template of jobTemplates) {
+      const resourceEntity: ResourceEntity =
+        this.buildApiEntityFromJobTemplate(template);
+      entities.push(resourceEntity);
+      this.logger.debug(`Discovered ResourceEntity ${template}`);
+    }
+
+    this.logger.info(`Applying the mutation with ${entities.length} entities`);
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: entities.map(entity => ({
+        entity,
+        locationKey: this.getProviderName(),
+      })),
+    });
+  }
+
+  private buildApiEntityFromJobTemplate(template: JobTemplate): ResourceEntity {
+    const location = `${this.baseUrl}/${template.url}`;
+    const jobTemplateName = template.name.replace(/ /g, '_');
+
+    return {
+      kind: 'Resource',
+      apiVersion: 'backstage.io/v1alpha1',
+      metadata: {
+        annotations: {
+          [ANNOTATION_LOCATION]: this.getProviderName(),
+          [ANNOTATION_ORIGIN_LOCATION]: this.getProviderName(),
+        },
+        name: `${jobTemplateName}`,
```

Comment:
I'm no AAP expert, but can we expect this to be unique? Can't you have different templates (maybe in different namespaces) with the same name? If this doesn't apply here, we should be fine. It just struck me as a potential issue.

Reply:
yeah i need to transform it , can't use the name as is